### PR TITLE
Hot mess changes!!!

### PR DESCRIPTION
**Pull request by SethPaul**

### CODE REVIEW

File: src/emoji.rs [hunk 1]
```diff
@@ -1,7 +1,32 @@
...
+            emoji_cache: HashMap::new(),
+        }
+    }
+
+    pub fn get_emoji(&mut self) -> &DynamicImage {
+        let index = random::<usize>() % EMOJIS_LEN;
+        self.get_emoji_by_index(index)
+    }
+
+    pub fn get_emoji_by_index(&mut self, index: usize) -> &DynamicImage {
```
Comment (TomPridham) [inline src/emoji.rs]:
```suggestion
    fn get_emoji_by_index(&mut self, index: usize) -> &DynamicImage {
```

Comment (TomPridham) [reply src/emoji.rs]: we don't want to expose an unsafe method without some guards around it

File: src/lib.rs [hunk 2]
```diff
@@ -1,70 +1,103 @@
...
-        .chunks_exact(3)
-        .zip(samples_b.chunks_exact(3))
-        .fold((0, 0, 0), |(r, g, b), (p_a, p_b)| {
-            (
-                r + (p_a[0] as i32 - p_b[0] as i32).abs(),
-                g + (p_a[1] as i32 - p_b[1] as i32).abs(),
-                b + (p_a[2] as i32 - p_b[2] as i32).abs(),
-            )
+        .chunks_exact(1)
+        .zip(samples_b.chunks_exact(1))
```
Comment (TomPridham) [inline src/lib.rs]:
```suggestion
        .zip(samples_b.iter())
```
i don't think you need to use `chunks_exact` if you are only taking one chunk

File: src/lib.rs [hunk 3]
```diff
@@ -1,70 +1,103 @@
...
 
-    for _ in 0..iterations {
+    let image_buffer_rgb = image_buffer.clone().to_rgb();
+    let (width, height) = image_buffer_rgb.dimensions();
+    let canvas_size = width * height;
+    let mut rng = thread_rng();
+    let mut new_img = DynamicImage::new_rgb16(width, height);
+    println!("image is {} by {} pixels", width, height);
+    let mut dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+    println!("dist is {}", dist);
```
Comment (TomPridham) [inline src/lib.rs]:
```suggestion
    let mut dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
```

File: src/lib.rs [hunk 4]
```diff
@@ -1,70 +1,103 @@
...
+
+    let mut placed_count = 0;
+    for _  in 0..canvas_size/20 {
         let e = emoji_cache.get_emoji();
-        let x: u32 = random::<u32>() % width;
-        let y: u32 = random::<u32>() % height;
+        let x: u32 = (1..width).choose(&mut rng).unwrap();
+        let y: u32 = (1..height).choose(&mut rng).unwrap();
```
Comment (TomPridham) [inline src/lib.rs]: i would maybe declare something like
`let width_range = 1..width` above, rather than doing this every iteration.
also, should it be `0..width`?

File: src/lib.rs [hunk 5]
```diff
@@ -1,70 +1,103 @@
...
+        overlay(&mut new_img, e, x, y);
+        placed_count= placed_count+1;
+    }
+
+    dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+    println!("dist is {}", dist);
```
Comment (TomPridham) [inline src/lib.rs]:
```suggestion
```

File: src/lib.rs (hunk 4 above)
Comment (SethPaul) [reply src/lib.rs]: I have been trying to do the single declare. 
So far I've only been able to do that if I clone anyway.
I've tried to see what is the most performant and so far this seems more performant in this loop than the random::<u32> method  
Working on it

File: src/lib.rs [hunk 6]
```diff
@@ -1,70 +1,103 @@
...
+        if index%1000==0 {
+            dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+            println!("iteration: {}, dist: {}, time: {}.{}, emoji: {}",
+                index,
+                dist,
+                now.elapsed().as_secs(),
+                now.elapsed().subsec_millis(),
+                placed_count
+                )
+        }
```
Comment (TomPridham) [inline src/lib.rs]: i would delete this and also maybe just delete `dist` itself since we're not really using it anymore

File: src/lib.rs [hunk 7]
```diff
@@ -1,70 +1,99 @@
...
+        if index%10000==0 && save_progress {
+            new_img.save(path)?;
         }
     }
+    println!("placed {} emoji", placed_count);
```
Comment (TomPridham) [inline src/lib.rs]:
```suggestion
```
it might be nice to return some stats if requested, but i think we could do that in another pr

File: src/lib.rs [hunk 8]
```diff
@@ -1,70 +1,80 @@
...
 
+fn subimage_compare(image_a: &DynamicImage, image_b: &DynamicImage, x: u32, y: u32 ) -> i64 {
+    let sub_image_a = image_a.crop_imm(x,y,16,16).to_rgb();
+    let sub_image_b = image_b.crop_imm(x,y,16,16).to_rgb();
+    measure_dist_chunks(&sub_image_a, &sub_image_b)
+}
+
 pub fn generate_image(
-    image_buffer: &[u8],
+    image_buffer: &DynamicImage,
```
Comment (TomPridham) [inline src/lib.rs:23]: this should get reverted, unless theres a reason to do it this way instead of accepting a byte array

Comment (TomPridham) [reply src/lib.rs:23]: mainly because im not confident using a `DynamicImage` will work in wasm. i'm gonna try and build this in wasm tonight and i'll report back

Comment (seth-paul-pdq) [reply src/lib.rs:23]: An image buffer is used for the crop method to get a sub image. Do you think that type in general is a problem or just as an input?

Comment (seth-paul-pdq) [reply src/lib.rs:23]: If it's just an input problem then we could convert after

Comment (seth-paul-pdq) [reply src/lib.rs:23]: If it's just an input problem then we could convert after

Comment (seth-paul-pdq) [reply src/lib.rs:23]: If it's just an input problem then we could convert after

Comment (TomPridham) [reply src/lib.rs:23]: i was thinking just as an input, but i looked into it and i don'
t think it will end up being an issue

**Review by TomPridham — APPROVED** [review]
